### ISSUE TITLE
Support primary/secondary failover mode using conntrackd and keepalived

### ISSUE THREAD

**Issue by AndrewGuenther** [AndrewGuenther/fck-nat#71]
The goal here is to use a primary/secondary configuration to improve fck-nat availability without the need for outside infrastructure. It's really important to me that fck-nat does not take on additional AWS service dependencies. The ideal state is that two autoscaling groups would allow for primary and secondary nodes to self-heal and as long as both nodes aren't offline at the same time failover would be seamless.

Here's a few articles documenting the approach in a non-AWS environment:

- https://satishdotpatel.github.io/ha-with-keepalived-and-conntrackd/
- https://geek.cm/2017/12/ha-network-gateway-with-keepalived-and-conntrackd/
- https://backreference.org/2013/04/03/firewall-ha-with-conntrackd-and-keepalived/index.html

Here's the rub: In AWS, we can't just willy nilly change our IP address. We have to explicitly move ENIs and EIPs. This takes some time. So the question becomes: In this configuration, can we move these resources fast enough to not incur downtime (packet loss is fine, downtime == dropped connections)?

This is going to take a lot of testing, but this is my ideal HA configuration for fck-nat 2.0 if it works.

**Comment by patrickdk77** [comment]
I've been running conntrackd + keepalived/pacemaker for firewall and then an ipvs pair with pacemaker as my failover for firewall/loadbalacing for 16 years now, it works well, just remember not to use multicast for things.
I haven't had much of a problem moving eip between systems in a useful time, though I haven't timed moving eni's

**Comment by AndrewGuenther** [comment]
@patrickdk77  Yeah, I've got _pretty good_ confidence this will work. I ran plenty of conntrackd+keepalived in ye olden days, but I'm not sure how well it'll translate to AWS networking.

**Comment by deepcrawltech** [comment]
Any progress on this work?

**Comment by RaJiska** [comment]
I have spent a truly indecent amount of time working on this and have finally been able to achieve true HA implemented with `keepalive` and `conntrackd` and implemented it in the Terraform module. Connections are maintained, and am even able to induce a failure mid-download of a large file, which is picked up right where it stopped when the ENI is moved to the passive instance.

There is however another problem which is the AWS Security Groups. Security Groups track connections independently of the instance, and as we are not able to fiddle with the state of the conntrack, it means that Security Groups on the backup instance will block previously established connections on the initial NAT following a failover as it has no knowledge of them (basically we'd need a security-group-conntrackd equivalent). The only way to make this work is to disable security groups connection tracking by allowing all inbound and all outbound ([documentation](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/security-group-connection-tracking.html)).

Due to these limitations, we will likely have to replace security groups with `iptables` as the effective firewall for NAT if we wish to have true HA. As managing both security groups and iptables configuration depending on the mode fck-nat is using (single instance, ASG, full HA) is likely a lot of hassle, I'd suggest entirely dropping security groups in favor of iptables for 2.0.

Any thoughts on this @AndrewGuenther ?

**Comment by quincycs** [comment]
Mentioning https://github.com/AndrewGuenther/fck-nat/issues/92 , as it seems to be a more promising true HA possibility.  

As an aside, I wanted to add my own story to support this functionality and put down my own words for the pain it solves.  Maybe it inspires and helps someone.

**Currently Fck-NAT doesn't support HA** in terms of preventing internet loss during a sudden EC2 availability issue.  Here's a screenshot of my AWS health issue that I encountered.  To clarify it wasn't an official AWS outage, instead a more isolated small scale issue that happened to impact me.  Lucky me.
https://www.reddit.com/r/aws/comments/1ojnkhr/why_hide_health_events/ 

The impact of my instance terminating resulted in ~23minutes of no internet connectivity effectively putting us offline.  Originally when testing this, launching instances was just ~5min or less... which I would have been totally fine with for such a rare scenario.  But for some unlucky reason the small scale issue also resulted in instances being slow to launch.  The ASG did auto fix the situation, so that's nice.  However it seems to me that the current fck-nat behavior isn't HA but instead automatic recovery.

**In comparison, the managed AWS NAT gateway offering claims to have redundancy within same AZ.**  

Quoting AWS documentation,
> Each NAT gateway is created in a specific Availability Zone and implemented with redundancy in that zone.
https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-basics.html

For now I am fighting with the business where we need to weigh the trade off of paying the silly $10K+ so I can have this silly amount of redundancy for the 23min of downtime.  Yeah - the managed NAT is going to cost us $10K+ a year while fck-nat costs us $420 (for the single instance).

**Comment by SikandAlex** [comment]
Coming in as a new contributor, but having read this thread end-to-end alongside #92 / #95, I wanted to put together the comparison @AndrewGuenther mentioned wanting to carve out time for. Posting here rather than opening a new issue because the context lives in this thread. Synthesis of existing discussions, not new opinions — just trying to make the decision easier.

## Baseline: what fck-nat ships today

Single-instance ASG with a floating ENI. The ASG replaces a failed instance; the replacement attaches the same ENI and traffic resumes. **In-flight connections die.** Recovery time is healthy-case ~30-90s; @quincycs's story above is the degraded-case anchor (23 minutes).

So both proposed paths answer the same question: **how do we preserve traffic during instance failure?**

## Path A — keepalived + conntrackd (this thread)

**Status:** Terraform module shipped by @RaJiska in Oct 2025. AMI-side wiring not yet done — keepalived and conntrackd aren't installed by the Packer build, and the fck-nat init script doesn't yet have role-discovery logic for active vs standby.

**How it works:** Two instances behind a floating ENI. keepalived runs VRRP between them; conntrackd replicates the kernel conntrack table from active → standby in near-real-time. On failover, the standby takes over the ENI and the replicated state lets in-flight flows continue.

**What works (per @RaJiska's testing in this thread):** connections survive failover — induced mid-download, picked up after the swap.

**Big caveat:** AWS Security Groups track connections independently from the instance. The standby's SG has no knowledge of flows established on the primary, so post-failover the SG drops the surviving connections regardless of what conntrackd does. The way around this is allowing all-inbound + all-outbound at the SG and letting iptables/nftables do the actual firewalling — which is the rationale behind @RaJiska's "drop SGs in favor of nftables for 2.0" proposal that @AndrewGuenther responded to positively above.

**Cost:** ~2× baseline instance cost. t4g.nano at ~\$3/mo each → ~\$6/mo for the pair.

**Open questions:**
- Conntrackd state replication has lag (typically a few ms). Flows established within that window won't survive a hard failure.
- VRRP on AWS needs to be unicast (per @patrickdk77's experience above).
- **No integration test today demonstrates "TCP connection survives failover" in CI.** That's the real verification gap before this is mergeable with confidence.

**Philosophical fit:** Zero AWS service dependencies. Matches the principle from this thread directly.

## Path B — GWLB (#92 / #95)

**Status:** PR #95 by @ozcrn was rebased onto 1.4 recently. @dbfreem load-tested and recommends merge. The remaining concern raised by @ddouglas is that GWLB build dependencies (boost and others) are baked into the AMI; should be built externally. @AndrewGuenther agreed on the thread and noted the same problem exists for NAT64's Jool DKMS install.

**How it works:** Many fck-nat instances behind a Gateway Load Balancer. GWLB distributes flows across targets via 5-tuple hashing. Active-active, scales horizontally with target count.

**What works:** Throughput scales linearly with instance count. Cost stays well below AWS managed NAT GW.

**Big caveat:** GWLB does not replicate flow state across targets. If a target dies, flows hashed to it die with it. The failure model is "remove unhealthy target, surviving flows on healthy targets continue, but flows on the dead target are lost." Different from Path A — no per-flow connection survival, but very clean degradation under partial failures.

**Cost:** GWLB endpoint roughly \$10-20/mo depending on usage + \$0.004/GB processed, plus N fck-nat instances. Still well below AWS managed NAT GW (~\$33-45/mo + \$0.045/GB processed).

**Open questions:**
- Build-deps externalization is the merge-blocker. Same infrastructure problem affects Jool/NAT64.
- GENEVE protocol handling in the fck-nat init script is the main code surface needing review.

**Philosophical fit:** One AWS service dependency (the GWLB endpoint). From this thread's framing this is a tension — but @AndrewGuenther's response on #92 ("I would love to see something like this as part of fck-nat") suggests the principle has flex when the cost/feature trade is right.

## Side-by-side

| Aspect | A: keepalived+conntrackd | B: GWLB |
| --- | --- | --- |
| AWS service deps | Zero | One (GWLB endpoint) |
| Scaling model | Active-passive, 2 instances | Active-active, N instances |
| In-flight connection survival on instance failure | Yes (conntrackd replication) | No (flow lost with dead target) |
| Cost above baseline | ~\$3-6/mo extra instance | ~\$10-20/mo GWLB + extra instances |
| AMI complexity added | keepalived + conntrackd + role-discovery logic | boost + GENEVE handling |
| Build-deps externalization required | No (apt/dnf packages) | Yes (blocker on #95) |
| Security-group implications | Forces SG → iptables/nftables firewall transition | None |
| Throughput ceiling | Single active instance | Scales with target count |
| Integration-test surface | Failover survival under load | Target health + flow distribution |

## Where the paths overlap

**Both benefit from externalizing native-build dependencies.** For Path A this is a non-issue (keepalived + conntrackd are apt/dnf packages). For Path B it's the merge-blocker. For NAT64 it's "should clean up." A shared external-build workflow (CI produces .rpm/.deb artifacts, Packer pulls prebuilt) unblocks B and improves A's and NAT64's AMI hygiene.

**Both interact with PR #128 (iptables → nftables).** If Path A goes forward and SGs come out as the connection-tracker, nftables becomes the actual firewall — making PR #128 load-bearing. If Path B goes forward, PR #128 is just modernization. Either way, PR #128 keeps its value.

## Combined approach (RaJiska's earlier suggestion)

"GWLB + per-target keepalived/conntrackd pair" — each GWLB target is itself a 2-instance HA pair. Gets you both scale and connection survival. Cost: ~2× target count + GWLB. Complexity: meaningful. Probably not the right first step but worth keeping in mind as a "v2 of the v1" option.

## Decision-aid framing

The trade-off comes down to which property fck-nat is optimizing for:

- **"Zero AWS deps + per-flow connection survival" → Path A.** Adds operational surface but aligns with the philosophy.
- **"Horizontal scale + active-active + simpler operational story" → Path B.** Trade: one AWS service dep, no flow-level survival.

If both were equally ready, Path A's philosophical alignment probably wins. But B's PR has been close to mergeable for longer — only the build-deps externalization remains as a real blocker — so the "fastest time-to-user-value" frame tilts toward B. The "most-asked-for user pain" frame (per @quincycs's report in this thread) tilts back toward A.

## Caveats on this synthesis

First-time contributor here; this is read across threads, not lived operations. Specifically:
- I haven't deployed either approach at production scale.
- Cost figures are approximate; regional variance applies.
- The conntrackd lag footprint and the GWLB target-removal latency both depend on workload and aren't well-quantified in what I've read.

## Concrete next-step offers if useful

- **For A:** I can take on the integration-test harness ("TCP connection survives a failover" demonstrated in CI). That's the verification gap.
- **For B:** the build-deps externalization could be done as a separate cross-cutting PR — it also benefits Jool/NAT64 and is the kind of foundation work that unblocks several stuck threads.
- **Either:** happy to defer to whichever direction you choose, and to wait if "decide first, then accept help" is the right sequence.
